Add minimum password strength

redshiftzero opened the pull request:
## Status

Ready for review

## Description of Changes

Steps towards #980 that in our internal engineering meeting on 3/9/17 we decided would be a good addition for 0.4 prior to merge of #1509. 

Changes proposed in this pull request:
* Adds a minimum password length requirement for new journalist users

## Testing

1. Provision dev VM
2. Attempt to make a journalist account with a weak password

## Deployment

Note that this PR does not require the reset of weak passwords that have already been set. We can also do that, but it should probably be done in a followup issue and PR, and I would argue _not_ for 0.4. 

## Checklist

### If you made changes to the app code:

- [x] Unit and functional tests pass on the development VM

conorsch commented:
Nice job on the tests. :+1: We should catch the exception in the CLI client, though:

```
./manage.py add-admin
Username: conordawg
Password:
Confirm Password:
Will this user be using a YubiKey [HOTP]? (y/N): n
['Traceback (most recent call last):\n', '  File "./manage.py", line 99, in _add_user\n    otp_secret=otp_secret)\n', '  File "<string>", line 4, in __init__\n', '  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/state.py", line 306, in _initialize_instance\n    manager.dispatch.init_failure(self, args, kwargs)\n', '  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__\n    compat.reraise(exc_type, exc_value, exc_tb)\n', '  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/state.py", line 303, in _initialize_instance\n    return manager.original_init(*mixed[1:], **kwargs)\n', '  File "/vagrant/securedrop/db.py", line 254, in __init__\n    self.set_password(password)\n', '  File "/vagrant/securedrop/db.py", line 286, in set_password\n    raise InvalidPasswordLength(password)\n', 'InvalidPasswordLength: Password needs to be at least 12 characters\n']
```

I'd rather see:

```
./manage.py add-admin
Username: conordawg
Password:
Confirm Password:
Error: Password needs to be at least 12 characters
```

with a non-zero exit code.

conorsch commented:
Also, I just got a "password too long" message in the Admin GUI when it should have been "password too short":

![password-too-long-eh](https://cloud.githubusercontent.com/assets/657862/25298085/1070e330-26a7-11e7-8017-b5407bfb3247.png)

I used a four-character password on that one. Confirmed that a too-long password still generates the too-long failure message.

fowlslegs commented:
This seems like a fine check to merge for now. I'm still in favor of merging https://github.com/freedomofpress/securedrop/pull/1509 or, better, eliminating passwords. I still don't think this is sufficient to stop users from picking an insecure password, but better than nothing.

redshiftzero commented:
Ah, good catches. Fixed the flashed messages in the webapp:

![screen shot 2017-04-21 at 3 41 01 pm](https://cloud.githubusercontent.com/assets/7832803/25298560/d205bf22-26aa-11e7-8b15-17e12cbac652.png)

![screen shot 2017-04-21 at 3 40 53 pm](https://cloud.githubusercontent.com/assets/7832803/25298563/d449c9a4-26aa-11e7-9401-0be300f7dda3.png)

and added a friendly error message for the CLI:

```
$ ./manage.py add-admin
Username: beep
Password: 
Confirm Password: 
Error: Password needs to be at least 12 characters.
```

conorsch commented:
```
vagrant@development:/vagrant/securedrop$ ./manage.py add-admin
Username: supdawg
Password:
Confirm Password:
Your password is too long (maximum length 128 characters). Please pick a shorter password.
```

Nice. :ok_hand: Also confirmed updates to the GUI notifications. LGTM!
